### PR TITLE
Loosen lifetime bounds on accessing named captures via `Index`.

### PR DESCRIPTION
**Pull request by jld**
The current definition equates the lifetime of the strings being
matched/captured and the string used as the capture name.  The latter
is usually `'static` (string literals), which leads to confusing
[borrow-check errors][err] on reasonable-looking code.

This change declares the two lifetimes as separate; this should be a
strict increase in the set of `impl` instances, so it's a minor version
bump.

[err][]: https://gist.github.com/jld/476a13a00ad05bd99a63

### CODE REVIEW

File: src/re.rs [hunk 1]
```diff
@@ -970,11 +970,11 @@ impl<'t> Index<usize> for Captures<'t> {
 ///
 /// # Panics
 /// If there is no group named by the given value.
-impl<'t> Index<&'t str> for Captures<'t> {
+impl<'t, 'i> Index<&'i str> for Captures<'t> {
```
Comment (BurntSushi) [inline src/re.rs]: Could you please add a bit of documentation here that briefly describes what the `'i` lifetime parameter stands for? Thanks! (The `'t` lifetime should also be documented, e.g., http://doc.rust-lang.org/regex/regex/struct.FindMatches.html)

**Comment by BurntSushi** [comment]
Can't believe I missed this! I have one doc nit, but otherwise looks good to me. Thank you!

File: src/re.rs (hunk 1 above)
Comment (jld) [inline src/re.rs]: Good point; thanks.

While I'm editing the docs: I noticed that the `str` can't outlive the `Captures` if it's obtained through either `Index` impl; should that be a separate PR?  It's a little surprising, so a warning and advice to use methods instead might be nice.  (This is part of the definition of the `Index` trait — normally `a[i]` is part of `a` and can't outlive it — so it can't be fixed without changing the `Output` type to `&'t str` and breaking compatibility.)

Comment (BurntSushi) [inline src/re.rs]: @jld So sorry I missed this comment! You raise a good point. Including those docs in this PR seems fine to me!

**Comment by jld** [comment]
Really sorry for dropping the ball on this.  (Also sorry for the rebase / force-push, but I had to fix the version bump.)  This version should take care of the doc comment issues.

File: Cargo.toml [hunk 2]
```diff
@@ -1,6 +1,6 @@
 [package]
 name = "regex"
-version = "0.1.48"  #:version
+version = "0.1.49"  #:version
```
Comment (BurntSushi) [inline Cargo.toml]: Could you remove the version bump please? I typically do them in separate commits with tags using a script.

File: src/re.rs [hunk 3]
```diff
@@ -951,6 +951,15 @@ impl<'t> Captures<'t> {
...
 /// Get a group by index.
 ///
+/// `'t` is the lifetime of the matched text.
+///
+/// The text can't outlive the `Captures` object if this method is
+/// used, because of how `Index` is defined (normally `a[i]` is part
+/// of `a` and can't outlive it); to do that, use [`at()`][]
+/// instead.
+///
+/// [`at()`]: #method.at
```
Comment (BurntSushi) [inline src/re.rs]: Could you remove the links here? I've generally avoided putting links to things in the docs because they're liable to become rotten.

File: src/re.rs [hunk 4]
```diff
@@ -965,13 +974,23 @@ impl<'t> Index<usize> for Captures<'t> {
...
 ///
+/// `'t` is the lifetime of the matched text and `'i` is the lifetime
+/// of the group name (the index).
+///
+/// The text can't outlive the `Captures` object if this method is
+/// used, because of how `Index` is defined (normally `a[i]` is part
+/// of `a` and can't outlive it); to do that, use [`name()`][]
+/// instead.
+///
+/// [`name()`]: #method.name
```
Comment (BurntSushi) [inline src/re.rs]: Same here for links.

**Comment by BurntSushi** [comment]
@jld OK, now I'm the one who has dropped the ball! I left a few more nits. If you like, I can pull your branch and make the changes myself---I feel bad about the back and forth!

**Comment by BurntSushi** [comment]
OK, I just went ahead and merged it in d8fb7f10a3c0010b0a364e969c586d1b72f8d14a. Thanks!

**Comment by BurntSushi** [comment]
Released on crates.io in `0.1.55`.
